### PR TITLE
fix: Adjust clipping logic to be closer to Windows

### LINKED ISSUE #7061: [WASM/Skia/Android] Elements being clipped within a `Grid` with a negative margin
### Current behavior

Place a child `Grid` inside of a parent `Grid`. With a negative margin on the child `Grid`, it should not be clipped by the parent `Grid`.

Observe that the child `Grid` is being improperly clipped:

image

### Expected behavior

WinUI behaviour:

image

### How to reproduce it (as minimally and precisely as possible)

```xml
<Grid Width="100"
	Height="100"
	Background="Purple">
	<Grid Margin="0, -4, -4, 0"
		Background="Yellow"
		VerticalAlignment="Top"
		HorizontalAlignment="Right">
		<Ellipse Height="30"
				Width="30" Fill="Red"/>
	</Grid>
</Grid>
```

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

Android, WebAssembly, Skia (GTK on Linux/macOS/Windows)

### IDE

Visual Studio 2019

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

Works on iOS

### LINKED ISSUE #9651: [WASM] `Canvas.Background` rendering not working correctly
### Current behavior

Please see attached: Canvas.Background should be filled completely - even 'outside' of what's visible at that point.

[Canvas.zip](https://github.com/unoplatform/uno/files/9443386/Canvas.zip)

### Expected behavior

WASM should work as WindowsAppSDK

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #10279: Content clipping by corner radius edge case
### Current behavior

Content is always clipped.

### Expected behavior

In some very weird edge cases, the content of an element "protrudes" from the border in UWP/WinUI

image

### How to reproduce it (as minimally and precisely as possible)

See above repro image from `Border_CornerRadius_BorderThickness` sample

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia, Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

WebAssembly, Android, iOS, macOS (AppKit), Mac Catalyst, Skia (WPF), Skia (GTK on Linux/macOS/Windows), Skia (Linux Framebuffer), Skia (Tizen)

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #13582: Clipping not applied properly when using RenderTransform
### Current behavior

```xml
	<Grid Width="100"
				Height="100" Background="Blue">
		<Ellipse Fill="Red"
					 Width="120"
					 Height="120">
			<Ellipse.RenderTransform>
				<RotateTransform Angle="45" />
			</Ellipse.RenderTransform>
		</Ellipse>
	</Grid>
```

image

### Expected behavior

Windows:

image

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

None

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

Skia. I haven't tested other platforms.

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #13590: UIElement_Transform_And_Clipping doesn't render correctly
On the left is current behavior. On the right is expected behavior (Windows)

image

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #7061, closes #10279, closes #9651, closes #13590, closes #13582

TODO: Verify https://github.com/unoplatform/uno/issues/7508, #8295

## PR Type

What kind of change does this PR introduce?

### <samp>🤖 Generated by Copilot at 29c9373</samp>

Fix element clipping logic in `InnerArrangeCore` method. Use WPF implementation as reference and handle different clipping scenarios based on element size and slot size.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Youssef1313** [comment]
Marking this PR as a draft to avoid accidental merge as this will require reviewing screenshot comparison first.

File: src/Uno.UI/UI/LayoutHelper.cs [hunk 1]
```diff
@@ -93,53 +93,41 @@ internal static Size GetMarginSize(this IFrameworkElement frameworkElement)
 		}
 
 		[Pure]
-		internal static (Point offset, bool overflow) GetAlignmentOffset(this IFrameworkElement e, Size clientSize, Size renderSize)
+		internal static Point GetAlignmentOffset(this IFrameworkElement e, Size clientSize, Size renderSize)
 		{
-			// Start with Bottom-Right alignment, multiply by 0/0.5/1 for Top-Left/Center/Bottom-Right alignment
```
Comment (Youssef1313) [inline src/Uno.UI/UI/LayoutHelper.cs:98]: The changes done here are to reduce unnecessary calculations. Previously, we would calculate the offset as bottom-right, then we throw away the calculation if we are aligning  horizontally to left or vertically to top. Now, the code is more straightforward.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Canvas_Measurement.cs [hunk 2]
```diff
@@ -134,6 +128,51 @@ public async Task When_Verify_Canvas_In_Canvas()
 			ImageAssert.HasColorAtChild(bitmap, clippedLocation, (float)clippedLocation.Width / 2, clippedLocation.Height / 2, Blue);
 		}
 
+		[TestMethod]
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Canvas_Measurement.cs:131]:
```suggestion
		[TestMethod]
#if __ANDROID__
		[Ignore("Fails on Android.")]
#endif
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Grid.cs [hunk 3]
```diff
@@ -309,6 +311,116 @@ public async Task When_Child_Added_Measure_And_Visible_Arrange()
 #endif
 		}
 
+		[TestMethod]
+		[RunsOnUIThread]
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Grid.cs:315]:
```suggestion
		[RunsOnUIThread]
#if __ANDROID__ || __IOS__
		[Ignore("Fails on Android and iOS.")]
#endif
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Grid.cs [hunk 4]
```diff
@@ -309,6 +311,116 @@ public async Task When_Child_Added_Measure_And_Visible_Arrange()
...
+
+			var yellowBounds = ImageAssert.GetColorBounds(bitmap, Colors.Yellow, tolerance: 5);
+			Assert.AreEqual(new Rect(new Point(84, 6), new Size(29, 29)), yellowBounds);
+
+			var redBounds = ImageAssert.GetColorBounds(bitmap, Colors.Red, tolerance: 5);
+			Assert.AreEqual(new Rect(new Point(84, 6), new Size(29, 29)), redBounds);
+		}
+
+		[TestMethod]
+		[RunsOnUIThread]
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_Grid.cs:376]:
```suggestion
		[RunsOnUIThread]
#if __ANDROID__ || __IOS__
		[Ignore("Fails on Android and iOS.")]
#endif
```

**Comment by Youssef1313** [comment]
Meaningful differences in screenshot compare:

- DoubleTappedTests [Bug fix]
- Transform_Ellipse_In_Border [Bug fix]
- Transform_Ellipse_in_two_grids [Regression]
- UIElement_Transform_And_Clipping [In CI, both screenshot are completely wrong, locally, what I see is that it was wrong and got fixed]
- TO BE CONTINUED

**Review by MartinZikmund — APPROVED** [review]

**Comment by MartinZikmund** [comment]
CC @jeromelaban, this seems very promising now!

**Review by dr1rrb — APPROVED** [review]
LGTM, @carldebilly might have an opinion

**Review by carldebilly — APPROVED** [review]

**Pull request merged**